cluster/gce/manifests/etcd.manifest

```diff
@@ -22,7 +22,7 @@
     "command": [
               "/bin/sh",
               "-c",
-              "if [ -e /usr/local/bin/migrate-if-needed.sh ]; then /usr/local/bin/migrate-if-needed.sh 1>>/var/log/etcd{{ suffix }}.log 2>&1; fi; exec /usr/local/bin/etcd --name etcd-{{ hostname }} --listen-peer-urls {{ etcd_protocol }}://127.0.0.1:{{ server_port }} --initial-advertise-peer-urls {{ etcd_protocol }}://{{ hostname }}:{{ server_port }} --advertise-client-urls http://127.0.0.1:{{ port }} --listen-client-urls http://127.0.0.1:{{ port }} {{ quota_bytes }} --data-dir /var/etcd/data{{ suffix }} --initial-cluster-state {{ cluster_state }} --initial-cluster {{ etcd_cluster }} {{ etcd_creds }} 1>>/var/log/etcd{{ suffix }}.log 2>&1"
+              "if [ -e /usr/local/bin/migrate-if-needed.sh ]; then /usr/local/bin/migrate-if-needed.sh 1>>/var/log/etcd{{ suffix }}.log 2>&1; fi; exec /usr/local/bin/etcd --name etcd-{{ hostname }} --listen-peer-urls {{ etcd_protocol }}://0.0.0.0:{{ server_port }} --initial-advertise-peer-urls {{ etcd_protocol }}://{{ hostname }}:{{ server_port }} --advertise-client-urls http://127.0.0.1:{{ port }} --listen-client-urls http://127.0.0.1:{{ port }} {{ quota_bytes }} --data-dir /var/etcd/data{{ suffix }} --initial-cluster-state {{ cluster_state }} --initial-cluster {{ etcd_cluster }} {{ etcd_creds }} 1>>/var/log/etcd{{ suffix }}.log 2>&1"
```

Comment:
does this expose the etcd data during the migration?

Reply:
also, this migration does not support multi-member etcd migrations, does it?

Reply:
I'm not sure 0.0.0.0 is actually what we want.

We previously had {hostname} here and I already had concerns in the original PR when changing that to 127.0.0.1 and apparently was right.

The correct way of fixing is is to revert to {{hostname}} here (or sth hostname-like) and start setting it in GKE zonal clusters to IP address (similarly like we do in regional clusters).

Reply:
@jpbetz - can you please take this?

Reply:
Alternatively we can keep existing behavior by replacing "0.0.0.0" by something like "(/usr/bin/host $hostname | head -1 | awk " .." )". This is what happening when hostname passed to etcd right now.  Plus it might be a good idea to add extra "127.0.0.1".

Reply:
I think it has to be very specific think in case of GKE regional clusters. It really need to be {{hostname}}

Reply:
My apologies, I should have checked the HA implications of the hostname->127.0.0.1 change more carefully when we upgraded to etcd 3.2.

`{{hostname}}` is no longer supported due to the etcd 3.2 upgrade we did recently. In etcd 3.1-, etcd would attempt to figure out which interface to bind to from the hostname in the  `--listen-peer-urls` URL.  This is not supported supported by etcd 3.2. Only IP addresses (and localhost) are allowed in the `--listen-peer-urls` URL. See https://coreos.com/etcd/docs/latest/op-guide/configuration.html for details.

For GCE the hostname resolves to the IP address of the eth0 network interface. The safest way to restore the previous behavior would be what @anjensan suggests and perform the hostname->IP lookup before passing the value to the flag.

Unfortunately, it looks like we can't rely on `host`, `nslookup` or `dig` being available on GCE hosts.

One option would modify [prepare-etcd-manifest](https://github.com/anjensan/kubernetes/blob/c6be41669117b89957f9f176c17a7606c7074617/cluster/gce/gci/configure-helper.sh#L1353) to add a allow the etcd host IP to be wired in, of default it to `hostname -i` if it is not provided. It would also need to be wired in via [gce/util.sh](https://github.com/kubernetes/kubernetes/blob/6235a60906bb4164d93ada1b9f98d05adeea8597/cluster/gce/util.sh#L779). To be 100% backward compatibile we should default it to the IP address of whatever `ETCD_HOSTNAME` resolves to, or, if it's an IP, it's value.

Reply:
Ok, since we have a dependency on python already, we do this:

`gce/gci/configure-helper.sh`:

```
function prepare-etcd-manifest {
  local host_name=${ETCD_HOSTNAME:-$(hostname -s)}
  local host_ip=$(python -c "import socket;print(socket.gethostbyname(\"$(host_name)\"))")
  ...
```

`cluster/gce/manifests/etcd.manifest`:

```
--listen-peer-urls {{ etcd_protocol }}://{{ host_ip}}:{{ server_port }}
```

Reply:
This won't work for regional clusters too. Because ETCD_HOSTNAME is actually set to real IP address (which is not IP address of the machine). And that's by design (I can share you a doc that describes it).

> {{hostname}} is no longer supported due to the etcd 3.2 upgrade we did recently

So the logic should be something like:
- if $ETCD_HOSTNAME is set and already is of the form of IP, use it. Otherwise use what you suggested.

Reply:
Some good news- Some testing of socket.gethostbename on gce and gke verified that if the input is an IP address, it is passed through without modification, so we don’t need to add logic for the case of $ETCD_HOSTNAME being an IP.

Reply:
If socket.gethostbyname returns the same IP if given with IP, the I'm fine with the approach.

Reply:
https://github.com/kubernetes/kubernetes/pull/60422